Various fixes and additions

LandonJerre opened the pull request:
- Fixed a crash, that happened when the user tried to delete multiple selected nodes and connections
- Fixed a crash, that happened when the user tried to load a scene from file into a non-empty scene
- Added functions to FlowScene for node position manipulation, and size readback. Before this, this data were only accessible through types that are not exported from the library, but on the other hand moving nodes in code is useful for automatic graph generation. (Like generating a basic graph that the user edits, instead of giving them an empty scene.)
- Added functions to Flowscene for iterating through the existing nodes. iterateOverNodeData just iterates through the nodes, passing the node data to the visitor lambda. iterateOverNodeDataDependentOrder iterates through the nodes in a specific order, it always visits a nodes all input nodes, before visiting the node itself. The latter is useful for codegeneration from the graph, where the order of the generated statements are important, and statements that feed other statements must be before them.
- Redid the context menu to be usable with a larger number of node types. Instead of dumping all the types into the menu, now all of them are put into a treewidget, under the categories that are passed during type registration. (A sideeffect is, that the filtering is noticably faster now, with the old way, filtering in the list on the picture froze the ui for a few moment after the first character inputed.) To illustrate the problem, and the solution:
![comp](https://cloud.githubusercontent.com/assets/7118075/23680380/508ce6ba-038b-11e7-9a24-19aa9cbb29c4.jpg)

LandonJerre commented:
The second point probably takes care of issue #54

paceholder commented on src/DataModelRegistry.hpp:
```diff
@@ -47,7 +50,7 @@ class NODE_EDITOR_PUBLIC DataModelRegistry
   template<typename ModelType, bool TypeConverter = false>
   void
-  registerModel(std::unique_ptr<ModelType> uniqueModel = std::make_unique<ModelType>())
+  registerModel(std::unique_ptr<ModelType> uniqueModel = std::make_unique<ModelType>(), const QString& category = "Nodes")
```
minor: please follow the style  `Type const &`

paceholder commented on src/FlowScene.cpp:
```diff
@@ -230,6 +230,98 @@ iterateOverNodes(std::function<void(Node*)> visitor)
+
+void
+FlowScene::
+iterateOverNodeDataDependentOrder(std::function<void(NodeDataModel*)> visitor)
```
I would simplify and restructure the first part as follows:

```
auto isNodeWithZeroInputs = [](Node const & node)
{
  bool result = true;

  auto model = node->nodeDataModel();

  for(std::size_t i = 0; i < model->nPorts(PortType::In); ++i)
  {
    auto connections = _node->nodeState().connections(PortType::In, i);

    if (!connections.empty())
    {
      result = false;
      break;
    }
  }

  return result;
};


std::size_t nodeIndex = 0;

for (auto const & idNodePair: _nodes)
{
  auto const &node = idNodePair.second;

  if (isNodeWithZeroInputs(node))
  {
    nodeIndexMap[node->id()] = nodeIndex;

    auto model = node->nodeDataModel();

    visitor(model);
  }

  ++nodeIndex;
}
```

paceholder commented on src/FlowScene.cpp:
```diff
@@ -230,6 +230,98 @@ iterateOverNodes(std::function<void(Node*)> visitor)
+        {
+          if (nodeIndexMap.find(conn.second->getNode(PortType::Out)->id()) == nodeIndexMap.end())
+          {
+            goto skip_node_dep;
```
I think it must be changed as above. It is possible to do the job without `goto`.

paceholder commented:
Thanks for this great add-on. I made a couple of comments

LandonJerre commented:
Fixed the code style issue (I made that mistake in another function prior, so fixed that as well), refactored iterateOverNodeDataDependentOrder to get rid of goto, plus I did a few minor simplifications on it, so hopefully it's much easier to read it now.

paceholder commented:
I will merge it. Please, next time, if possible do several independent pull-requests or commits for each implemented feature.

LandonJerre commented:
Sorry about that, I realized after I finished this, that I should've separated my changes. All the stuff I did/am doing right now are separated logically, so the next pull requests won't be this messy hopefully.
